### PR TITLE
[PTCD-636] Switch delete to soft delete (aka archive)

### PR DESCRIPTION
**Pull request by timabell**
[PTCD-636](https://skillsfundingagency.atlassian.net/browse/PTCD-636)

# What?

* Convert the two delete endpoints to use archive instead

# Why?

We *think* that hard delete is causing changes to not show up in the CosmosDB Change Feed Listener (CFL), and as such result in data inconsistency and errors across the system.

# What no tests?

There are no tests in the patch for this change in behaviour because the existing system architecture has made this virtually impossible without major changes. This has been discussed at length with the team and we are resigned to moving forward without test coverage in this instance.

# Course / CourseRun statuses

I'm informed by fellow devs that course status is calculated based on aggregation of course run statuses, so shouldn't be any need to set the status on the actual course. Ideally we'd wnat to clean up the models to prevent inconsistent states, but given the state of the project that will be a big task.

The code currently makes slightly strange use of the two statuses but it's a minimal change given lack of test coverage.

# Manual testing

This has been tested by finding a provider with courses in all the affected statuses, running each function and checking the resultant state of the data. Seems okay.

## Queries
```
-- run+course status counts
SELECT
	c.CourseStatus, r.RecordStatus CourseRun_Status, count(r.id) qty
FROM c
	join r in c.CourseRuns
where c.ProviderUKPRN = 10022802
group by r.RecordStatus, c.CourseStatus
```
```
-- course status counts
SELECT c.CourseStatus, count(c.id) qty
FROM courses c
where c.ProviderUKPRN = 10022802
group by c.CourseStatus
```

## Before running the funcs
```
[
    {
        "CourseStatus": 32,
        "CourseRun_Status": 32,
        "qty": 4
    },
    {
        "CourseStatus": 16,
        "CourseRun_Status": 16,
        "qty": 10
    },
    {
        "CourseStatus": 1,
        "CourseRun_Status": 1,
        "qty": 20
    },
    {
        "CourseStatus": 4,
        "CourseRun_Status": 4,
        "qty": 680
    }
]
```
```
[
    {
        "CourseStatus": 32,
        "qty": 4
    },
    {
        "CourseStatus": 16,
        "qty": 10
    },
    {
        "CourseStatus": 1,
        "qty": 19
    },
    {
        "CourseStatus": 4,
        "qty": 293
    }
]
```
## After bulk upload delete
http://localhost:7071/api/DeleteBulkUploadCourses?ukprn=10022802

```
[
    {
        "CourseStatus": 1,
        "qty": 19
    },
    {
        "CourseStatus": 4,
        "qty": 307
    }
]
```
## After delete live
http://localhost:7071/api/DeleteCoursesByUKPRN?ukprn=10022802
```
[
    {
        "CourseStatus": 4,
        "qty": 326
    }
]
```

### CODE REVIEW

File: src/Dfc.ProviderPortal.Courses/Services/CoursesService.cs
Comment (gunndabad) [inline src/Dfc.ProviderPortal.Courses/Services/CoursesService.cs:334]: This predicate would be better done in the Cosmos query

Comment (timabell) [reply src/Dfc.ProviderPortal.Courses/Services/CoursesService.cs:334]: Yeah I agree; I'm trying to make the minimal possible change to this test-less production code.
Do you want to suggest a patch, or discuss further in slack/video?

**Comment by timabell** [comment]
rebased

**Review by gunndabad — APPROVED** [review]

**Comment by timabell** [comment]
[Release-228](https://sfa-gov-uk.visualstudio.com/Digital%20First%20Careers/_releaseProgress?releaseId=17059&_a=release-pipeline-progress) for testing

**Review by soharadesign — APPROVED** [review]

**Comment by timabell** [comment]
Caught up with Hari (QA).
Unsurprisingly the state of the production data causes some unwanted behaviour. On hold till we figure this out.

**Comment by timabell** [comment]
Blocked until #171 is merged

**Comment by timabell** [comment]
Also blocked by https://github.com/SkillsFundingAgency/dfc-coursedirectory/pull/1732

**Review by miriyalan — APPROVED** [review]

**Pull request merged**
